data/transactions/logic/eval.go

```diff
@@ -3372,11 +3579,28 @@ func opAppLocalPut(cx *EvalContext) {
 		return
 	}
 
-	addr, accountIdx, err := cx.accountReference(cx.stack[pprev])
-	if err == nil {
-		err = cx.Ledger.SetLocal(addr, key, sv.toTealValue(), accountIdx)
+	addr, accountIdx, err := cx.mutableAccountReference(cx.stack[pprev])
+	if err != nil {
+		cx.err = err
+		return
 	}
 
+	// if writing the same value, do nothing, matching ledger behavior with
+	// previous BuildEvalDelta mechanism
+	etv, ok, err := cx.Ledger.GetLocal(addr, cx.appID, key, accountIdx)
+	if err != nil {
+		cx.err = err
+		return
+	}
+
+	tv := sv.toTealValue()
+	if !ok || tv != etv {
+		if _, ok := cx.Txn.EvalDelta.LocalDeltas[accountIdx]; !ok {
+			cx.Txn.EvalDelta.LocalDeltas[accountIdx] = basics.StateDelta{}
+		}
+		cx.Txn.EvalDelta.LocalDeltas[accountIdx][key] = tv.ToValueDelta()
```

Comment:
Shouldn't the size of `LocalDeltas` be capped to `config.MaxEvalDeltaAccounts` ?

Reply:
That's an interesting question, and something I'm still grappling with for allocbounds.  I don't think that our code explicitly enforces many of the allocbounds.  Instead, the allocbounds are calculated so that they can not be exceeded (they are usually the max across _any_ protocol version, or something even more conservative).

I do think we have a problem if we ever get that wrong (as we've been talking about the last few days). But I don't know if should just sprinkle checks around.

I suppose I would feel safer if I wrote some sort of `func (ad ApplyData) checkAllocBounds()` that recursively checked them, and invoked it at the end of evaluation.  Does that sound reasonable?

Reply:
The high level design was to write the code so that it wouldn't have the ability to generate unrealistic sizes of  arrays/maps. Then, the msgp allocbound were set to align with these.

I don't believe that we have too many locations that have these issue. I think that it would be less error prone to perform that inline rather than "export" that process. ( i.e. keep the limits next to the variable )

Reply:
In that regard, this is true here.  The allocbound is huge (2048) but the number of different accounts that can be the accountIdx here is limited to 5. (The number of accounts in the static array, plus one for sender.)